### DISCUSSION TITLE
Basecamp API Library

### DISCUSSION THREAD

**Discussion by peterjohnhunt** [johnlindquist/kit#148] [Share]
This is very much still WIP, but i've been building a library for the [Basecamp 3 api](https://github.com/basecamp/bc3-api) so i can build a bunch of automations against that.

Here is a gist of the lib file:

https://gist.github.com/peterjohnhunt/128ded029c93fb8f6c4845df7fc590c2

and here is an example of how it could be used to create a task

```
// Menu: Basecamp - Create Task
// Description: Create a Basecamp task

let { me, selectTodoList, createTodo } = await lib("basecamp")

let list = await selectTodoList()

setHint(`${list.bucket.name} - ${list.title}`)

let title = await arg('Task name:')

let today = new Date()
let dd = String(today.getDate()).padStart(2, '0')
let mm = String(today.getMonth() + 1).padStart(2, '0')
let yyyy = today.getFullYear()

let response = await createTodo({
  content: title,
  assignee_ids: [me.id],
  due_on: `${yyyy}-${mm}-${dd}`
}, list.bucket.id, list.id)

setHint(`${title} Created!`)

wait(2000)
```

**Comment by peterjohnhunt** [comment]
I'm super excited for a beta to release with the drop and hotkey functions and how i might use those with this as well!

**Reply by peterjohnhunt** [reply]
ANNND just saw beta 94 was deployed! excited to try it out. This is the most fun i've had in a long time! :D

**Reply by johnlindquist** [reply]
Sorry, I had to pull the release. Found a minor bug around opening files after creating a new script that I'm hunting down. Wish me luck 🙂

**Reply by johnlindquist** [reply]
@peterjohnhunt Here you go! Party time: https://github.com/johnlindquist/kit/discussions/150

**Reply by peterjohnhunt** [reply]
Aww snap! I know what i'll be doing all night tonight! 😄

**Comment by cfritze64** [comment]
super helpful as an example of a user's library. I've been having a little trouble sorting out the overall structure of the directories: to access this from a script, where do you save this library file? thx

**Reply by johnlindquist** [reply]
Any .js file you put in the `lib` dir is accessible by `await lib("foo")`

<img width="583" alt="Screen Shot 2021-04-20 at 3 28 26 PM" src="https://user-images.githubusercontent.com/36073/115466269-16fe0780-a1ed-11eb-9d32-8b985c6090c3.png">

**Comment by cfritze64** [comment]
Am I making a basic javascript error, or misunderstanding some part of scriptkit's structure?

I've got this code in a file in the correct lib folder:

export let dummy = async(aVar) => {
		let result = await applescript(String.raw `
		 return ${aVar}+1`);
		 return result;
	 }

(this is obviously dummed down from a more real AppleScript operation. Whatever complex AppleScript call I put in dummy gets executed faithfully.) 

My script makes the following call:

let result = dummy(1);
  inspect(result);

The script loads dummy fine. Other functions that I import from that lib file (that don't return a result) operate as expected. But dummy only ever gets me an empty object if I show() or inspect() the result.

**Reply by johnlindquist** [reply]
I see that `dummy` is an async function.

You'll need to use `await dummy()` or you'll just get a Promise back. (Any JS fn marked with `async` returns a Promise)

**Comment by cfritze64** [comment]
I knew when I got the answer it'd be a forehead-slapper. Doh! I've got scripts like this where I have 5 or 6 function calls piled up in my script, none of them including "await" in the call , none of them expect a result, and that works great, presumably because the AppleScript code I'm calling runs asynchronously and I'm never waiting on a promise to fill a variable to be returned, is that it? Many, many thanks!

chr.

**Reply by johnlindquist** [reply]
Yup, sounds right. `await` is just like passing a callback to `then(callback)`. If you don't pass a callback, the code still runs, but there's no way for you to handle the result.

**Comment by peterjohnhunt** [comment]
### Update on how i'm using it:
I have built a [number of small scripts](https://github.com/peterjohnhunt/kit-scripts/tree/master/scripts) around the BaseCamp API and have them all separated utilizing a [BaseCamp API library](https://github.com/peterjohnhunt/kit-scripts/blob/master/lib/basecamp.js)

I have created a small helper function to take a list of scripts and turn them into choices / allow them to be run when selected:

```
export const scriptChoices = async (options, placeholder = 'Select a script') => {
  let script = await arg(placeholder, options)
  
  await run(script)
}
```

### Current Usage
Using that helper, I have broken the scripts into three main categories: Shortcuts, Messages, and Tasks and built a basecamp main file that lists out the scripts in separate tabs:

```
// Menu: Basecamp
// Description: Command Menu
// Shortcut: shift alt b

let { scriptChoices } = await lib("helpers")
let { me } = await lib("basecamp")

let placeholder = `Hello ${me.name}:`

onTab('Shortcuts', () => scriptChoices([{
  name: 'Create Lead',
  value: 'basecame-create-lead'
},{
  name: 'Create Reminder',
  value: 'basecame-create-reminder'
}], placeholder))

onTab('Tasks', () => scriptChoices([{
  name: 'Create Task',
  value: 'basecamp-create-task'
},{
  name: 'View Task',
  value: 'basecamp-view-task'
}], placeholder))

onTab('Messages', () => scriptChoices([{
  name: 'Create Message',
  value: 'basecamp-create-message'
},{
  name: 'View Message',
  value: 'basecamp-view-message'
}], placeholder))
```

I have also created scripts for each category so they can have their own shortcut/entry point:

```
// Menu: Basecamp | Task
// Description: Task Commands

let { scriptChoices } = await lib("helpers")
let { me } = await lib("basecamp")

scriptChoices([{
  name: 'Create Task',
  value: 'basecamp-create-task'
},{
  name: 'View Task',
  value: 'basecamp-view-task'
}], `Hello ${me.name}:`)
```

I may eventually change the "main" script to be something more like: `onTab('Tasks', () => run('basecamp-task'))` but the tabs get removed when that tab is selected since it essentially starts a new run all together which kinda makes sense since it'd get a little wonky if there was a script with tabs that called another script with tabs.
